File: Algorithms/p145_Binary_Tree_Postorder_Traversal/p145_Binary_Tree_Postorder_Traversal_2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        stack = []
        cur = (root,0) # [1]: childs visited
        while True:
            if cur[0]:
                if not cur[1]:
                    stack.append((cur[0],1))
                    stack.append((cur[0].right,0))
                    cur = (cur[0].left,0)
                else:
                    res.append(cur[0].val)
                    cur = (None,1)
            else:
                if stack:
                    cur = stack.pop()
                else:
                    break
        return res

File: Algorithms/p145_Binary_Tree_Postorder_Traversal/p145_Binary_Tree_Postorder_Traversal_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        typedef pair<TreeNode*,bool> TreeNodePair; // second: childs visited
        vector<int> res;
        vector<TreeNodePair> stack;
        TreeNodePair cur(root,0);
        while(1){
            if(cur.first){
                if(!cur.second){
                    stack.push_back(TreeNodePair(cur.first,1));
                    stack.push_back(TreeNodePair(cur.first->right,0));
                    cur = TreeNodePair(cur.first->left,0);
                }else{
                    res.push_back(cur.first->val);
                    cur.first = 0;
                }
            }else{
                if(!stack.empty()){
                    cur = stack.back();
                    stack.pop_back();
                }else{
                    break;
                }
            }
        }
        return res;
    }
};
